src/Common/tests/System/Net/WebSockets/WebSocketCreateTest.cs

```diff
@@ -178,7 +150,117 @@ public async Task ReceiveAsync_ServerSplitHeader_ValidDataReceived()
+                        Assert.Equal(WebSocketState.Open, socket.State);
+
+                        Assert.True(stream.CanRead);
+                        Assert.True(stream.CanWrite);
+
+                        if (explicitCloseAsync) // make sure CloseAsync ends up disposing the stream
+                        {
+                            await socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
+                            Assert.False(stream.CanRead);
+                            Assert.False(stream.CanWrite);
+                        }
+                    }
+
+                    Assert.False(stream.CanRead);
+                    Assert.False(stream.CanWrite);
+                }
+            }
+        }
+
+        [OuterLoop] // Connects to external server.
+        [Theory]
+        [MemberData(nameof(EchoServersAndBoolean))]
+        public async Task WebSocketProtocol_CreateFromConnectedStream_CloseAsyncAfterCloseReceivedClosesStream(Uri echoUri, bool useCloseOutputAsync)
+        {
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                bool secure = echoUri.Scheme == "wss";
+                client.Connect(echoUri.Host, secure ? 443 : 80);
+
+                using (Stream stream = await CreateWebSocketStream(echoUri, client, secure))
+                using (WebSocket socket = CreateFromStream(stream, false, null, TimeSpan.FromSeconds(10)))
+                {
+                    Assert.NotNull(socket);
+                    Assert.Equal(WebSocketState.Open, socket.State);
+
+                    // Ask server to send us a close
+                    await socket.SendAsync(Encoding.UTF8.GetBytes(".close"), WebSocketMessageType.Text, true, default);
+
+                    // Verify received server-initiated close message.
+                    WebSocketReceiveResult recvResult = await socket.ReceiveAsync(new byte[256], default);
+                    Assert.Equal(WebSocketCloseStatus.NormalClosure, recvResult.CloseStatus);
+                    Assert.Equal(WebSocketCloseStatus.NormalClosure, socket.CloseStatus);
+                    Assert.Equal(WebSocketState.CloseReceived, socket.State);
+
+                    Assert.True(stream.CanRead);
+                    Assert.True(stream.CanWrite);
+
+                    await (useCloseOutputAsync ?
+                        socket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None) :
+                        socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None));
+
+                    Assert.False(stream.CanRead);
+                    Assert.False(stream.CanWrite);
+                }
+            }
+        }
+
+        private static async Task<Stream> CreateWebSocketStream(Uri echoUri, Socket client, bool secure)
+        {
+            Stream stream = new NetworkStream(client, ownsSocket: false);
```

Comment:
nit: can use 'var'.

Reply:
This one can't, actually.  The variable needs to be typed as `Stream` rather than `NetworkStream` so that the `SslStream` can be stored into it.